UnitySDK/Assets/ML-Agents/Scripts/InferenceBrain/MetaDataLoader.cs

```diff
@@ -0,0 +1,485 @@
+                        Shape = new long[]
+                        {
+                            -1, _brainParameters.vectorActionSize.Sum()
+                        },
+                        ValueType = Tensor.TensorType.FloatingPoint,
+                        Data = null
+                    });
+            }
+            var memory = GetIntScalar(TensorNames.MemorySize);
+            if (memory > 0)
+            {
+                tensorList.Add(new Tensor()
+                {
+                    Name = TensorNames.RecurrentOutput,
+                    Shape = new long[2]
+                    {
+                        -1, memory
+                    },
+                    ValueType = Tensor.TensorType.FloatingPoint,
+                    Data = null
+                });
+            }
+
+            return tensorList;
+        }
+
+
+        /// <summary>
+        /// Queries the InferenceEngine for the value of a variable in the graph given its name.
+        /// Only works with int32 Tensors with zero dimensions containing a unique element.
+        /// If the node was not found or could not be retrieved, the value -1 will be returned. 
+        /// </summary>
+        /// <param name="name">The name of the Tensor variable</param>
+        /// <returns></returns>
+        private int GetIntScalar(string name)
+        {
+            var outputs = new Tensor[]
+            {
+                new Tensor()
+                {
+                    Name = name,
+                    ValueType = Tensor.TensorType.Integer,
+                    Shape = new long[] { },
+                    Data = new long[1]
+                },
+            };
+            try
+            {
+                _engine.ExecuteGraph(new Tensor[0], outputs);
+            }
+            catch (Exception e)
+            {
+                Debug.Log("Node not in graph: " + name + ". The following error occured : \n" + e);
+                return -1;
+            }
+
+            return (outputs[0].Data as int[])[0];
+        }
+
+        /// <summary>
+        /// Retrieves an IEnumerable of string corresponding to the failed compatibility checks
+        /// between the InferenceEngine and the BrainParameters. 
+        /// </summary>
+        public IEnumerable<string> GetChecks()
+        {
+            return _failedModelChecks;
+        }
+
+        /// <summary>
+        /// Generates the list of failed checks that failed when comparing the data from the Model
+        /// and from the BrainParameters
+        /// </summary>
+        private void GenerateChecks()
+        {
+            _failedModelChecks.Clear();
+            if (_engine == null)
+            {
+                _failedModelChecks.Add(
+                    "There is no model for this Brain, cannot run inference. " +
+                    "(But can still train)");
+                return;
+            }
+
+            var modelApiVersion = GetIntScalar(TensorNames.VersionNumber);
```

Comment:
Lines 147-150 and CheckIntScalarPresenceHelper are a bit problematic. For adding a new scalar, or removing one, you have to change in 3 places whereas essentially it's copy paste. I'd recommend here creating 
List<string> RequiredScalarFields = ... { Version, Memory, ... }
foreach (var field in Required...)
if field == -1
then add check bla bla

Looks prettier, no copy pasta, and if you need to add/remove a field it's only in one clear place.

modelApiVersion is special, so you'll need a second check for it, but still you'll remove 90% of the copy paste.